### PR TITLE
Add ./x.py check src/{libstd,libtest,librustc}

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
This currently only supports a limited subset of the full compilation,
but is likely 90% of what people will want and is possible without
building a full compiler (and also building LLVM). In theory, this means
that contributors who don't want to build LLVM now have an easy way to
compile locally, though running tests won't work.

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
cc @rust-lang/compiler

r? @alexcrichton

File: src/bootstrap/flags.rs [hunk 1]
```diff
@@ -210,6 +215,21 @@ Arguments:
...
+    This subcommand accepts a number of paths to directories to the crates
+    and/or artifacts to compile. For example:
+
+        ./x.py check src/libcore
+        ./x.py check src/libcore src/libproc_macro
+
+    If no arguments are passed then the complete artifacts are compiled: std, test, and rustc. Note
+    also that since we use `cargo check`, by default this will automatically enable incremental
+    compilation, so there's no need to pass it separately, though it won't hurt. We also completely
+    ignore the stage passed, as there's no way to compile in non-stage 0 without actually building
```
Comment (kennytm) [inline src/bootstrap/flags.rs]: Checking in non-stage-0 may still be useful if we want to check libstd after some new semantic is implemented:

```sh
./x.py build --stage 0 src/librustc
./x.py check --keep-stage 1 --stage 1 src/libstd
```

**Comment by michaelwoerister** [comment]
What exactly do these do?

**Comment by Mark-Simulacrum** [comment]
This runs `cargo check` instead of `cargo build` which nets much faster builds. Currently only works with stage 0, though we could technically expand it to work with any stage.

**Comment by michaelwoerister** [comment]
OK, that's great!

**Comment by alexcrichton** [comment]
Oops sorry about that, looks great to me!

r=me

**Comment by matklad** [comment]
Should this option be mentioned in https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#the-build-system?

**Comment by Mark-Simulacrum** [comment]
Probably, yes. I'll open an issue.

**Pull request merged**